File: rust/wallpaper-tui/src/fx.rs
//! Animation overlay for the wallpaper TUI: a re-triggerable preview
//! crossfade. All motion is opt-in and collapses to a no-op when
//! `DOTS_NO_ANIM` is set; the caller also gates on terminal capabilities
//! (`use_caps`) so raw VTs degrade to instant cuts.
//!
//! The crossfade is a one-shot 0→1 opacity ramp re-triggered on every
//! selection change. The view blends the new preview `Bitmap` toward the pane
//! background by the cached opacity (see [`blend_bitmap`]) — the `Image` widget
//! itself paints opaque mosaic cells, so the fade happens at the pixel level
//! before the mosaic renderer. The free helpers [`crossfade_curve`] /
//! [`blend_bitmap`] expose the math to unit tests via a deterministic
//! `Clock::fixed`.

use std::time::Duration;

use abstracttui::anim::{Clock, Easing, Tween};
use abstracttui::base::Rgba;
use abstracttui::gfx::Bitmap;
use abstracttui::reactive::request_frame;

/// True unless `DOTS_NO_ANIM` is set.
#[must_use]
pub fn animations_enabled() -> bool {
    std::env::var("DOTS_NO_ANIM").is_err()
}

const CROSSFADE_DUR: Duration = Duration::from_millis(150);

/// Live, re-triggerable preview crossfade. Held in a `Signal<Fx>` in the root
/// scope.
///
/// The opacity ramp is advanced once per frame by [`Fx::tick`] (called from the
/// app loop) and the eased result cached in `opacity`. The view reads that
/// cache via [`Fx::crossfade_opacity`] — it never mutates the signal during
/// render, which keeps the reactive damage contract clean (no write-during-read
/// re-render storms).
#[derive(Clone)]
pub struct Fx {
    clock: Clock,
    crossfade_start: Option<Duration>,
    /// Cached eased crossfade opacity (0.0..=1.0), refreshed by `tick`.
    opacity: f32,
}

impl Fx {
    /// New overlay on the given (real, in production) clock.
    #[must_use]
    pub fn new(clock: Clock) -> Self {
        Self {
            clock,
            crossfade_start: None,
            opacity: 1.0,
        }
    }

    /// Re-trigger the preview crossfade (called on selection change). A no-op
    /// when animations are disabled.
    pub fn retarget_crossfade(&mut self) {
        if !animations_enabled() {
            return;
        }
        self.retarget_crossfade_force();
    }

    /// Ungated crossfade re-trigger — the primitive `retarget_crossfade`
    /// delegates to. Tests use this directly so they don't race with the
    /// env-mutating `animations_enabled` test (tests run in parallel threads
    /// sharing one process env).
    pub fn retarget_crossfade_force(&mut self) {
        self.crossfade_start = Some(self.clock.now());
        request_frame();
    }

    /// Advance the ramp to the clock's `now` and cache the eased opacity. The
    /// app loop calls this once per frame; tests drive it via [`Fx::advance`] on
    /// a `Clock::fixed`. Re-requests a frame while the fade is in flight so the
    /// loop keeps pumping until it settles.
    pub fn tick(&mut self) {
        let now = self.clock.now();
        self.opacity = self.opacity_at(now);
        if self.crossfade_start.is_some() {
            request_frame();
        }
    }

    /// Advance the clock by `dur` then `tick`. The deterministic tests use this
    /// with `Clock::fixed` (the real loop uses `Clock::real` + `tick`).
    pub fn advance(&mut self, dur: Duration) {
        self.clock.advance(dur);
        self.tick();
    }

    /// Cached eased crossfade opacity (0.0..=1.0). Peek, no mutation.
    #[must_use]
    pub fn crossfade_opacity(&self) -> f32 {
        self.opacity
    }

    /// Eased opacity at `now`; clears the one-shot once `CROSSFADE_DUR` elapses
    /// (leaving opacity pinned at 1.0). Internal helper for `tick`.
    fn opacity_at(&mut self, now: Duration) -> f32 {
        let Some(start) = self.crossfade_start else {
            return 1.0;
        };
        let elapsed = now.saturating_sub(start);
        if elapsed >= CROSSFADE_DUR {
            self.crossfade_start = None;
            return 1.0;
        }
        // EaseOut: front-loads the fade so the new preview reads fast, then
        // settles gently — matches the installer's progress easing.
        Tween::new(0.0_f32, 1.0, CROSSFADE_DUR)
            .with_easing(Easing::EaseOut)
            .sample(elapsed)
            .clamp(0.0, 1.0)
    }
}

/// Blend `bmp` toward `bg` by `opacity` (0.0 → fully `bg`, 1.0 → fully `bmp`),
/// returning a fresh `Bitmap`. Used by the view to render the crossfade: feed
/// the blended bitmap to `Image::from_bitmap` while [`Fx::crossfade_opacity`]
/// is in flight. Cheap for a thumbnail (a few thousand px) over a ~150 ms fade.
#[must_use]
pub fn blend_bitmap(bmp: &Bitmap, bg: Rgba, opacity: f32) -> Bitmap {
    let t = opacity.clamp(0.0, 1.0);
    let mut out = Bitmap::new(bmp.width(), bmp.height(), bg);
    for (dst, src) in out.pixels_mut().iter_mut().zip(bmp.pixels()) {
        *dst = bg.lerp(*src, t);
    }
    out
}

/// Deterministic probe over the crossfade `Tween`, sampled at arbitrary times
/// (the `Tween` is stateless, so call order does not matter). Used by tests.
pub struct CrossfadeProbe {
    t: Tween<f32>,
}

impl CrossfadeProbe {
    /// Eased opacity at `ms` milliseconds into the fade.
    #[must_use]
    pub fn now(&self, ms: u64) -> f32 {
        self.t.sample(Duration::from_millis(ms))
    }
}

/// Eased 0→1 over `dur_ms` with `EaseOut`. The property tested: 0 at the start,
/// 1 at the duration, strictly between mid-fade.
#[must_use]
pub fn crossfade_curve(_clock: Clock, dur_ms: u64) -> CrossfadeProbe {
    CrossfadeProbe {
        t: Tween::new(0.0_f32, 1.0, Duration::from_millis(dur_ms)).with_easing(Easing::EaseOut),
    }
}

File: haskell/wallpaper-tui/src/WallpaperTui/Fx.hs
-- | Animation overlay for the wallpaper TUI: a re-triggerable preview
-- crossfade. Mirrors @rust/wallpaper-tui/src/fx.rs@. All motion is opt-in and
-- collapses to a no-op when @DOTS_NO_ANIM@ is set.
--
-- The crossfade is a one-shot 0→1 opacity ramp re-triggered on every
-- selection change. The view blends the new preview 'Bitmap' toward the pane
-- background by the cached opacity ('blendBitmap').
--
-- 'Fx' is PURE: it carries its own time field ('fxTime') rather than an
-- abstracttui 'Clock' (which is IO), so 'fxTick'\/'fxAdvance'\/'retarget' are
-- pure 'Fx -> Fx' functions — matching the Rust API and the
-- 'sigUpdate'\/'sigUpdateIO' pure-function contract. The production loop
-- seeds real time via 'fxTickAt' (passing @clockNow@ each frame); the tests
-- drive time directly via 'fxAdvance' on a zero-start 'fxNew' (matching
-- @Clock::fixed()@).
module WallpaperTui.Fx
  ( Fx (..)
  , fxNew
  , fxTick
  , fxTickAt
  , fxAdvance
  , retargetCrossfade
  , retargetCrossfadeForce
  , crossfadeOpacity
  , crossfadeDurMs
  , animationsEnabled
  , CrossfadeProbe (..)
  , crossfadeCurve
  , blendBitmap
  ) where

import Data.Word (Word8)
import System.Environment (lookupEnv)

import AbstractTUI.Base.Color (Rgba (..))
import AbstractTUI.Gfx.Bitmap (Bitmap (..), mapPixels)
import AbstractTUI.Anim (Tween, tween, tweenEasing, tweenSample, Easing (EaseOut))

import WallpaperTui.Accent (roundHalfAway)

-- | The crossfade duration in milliseconds (mirrors @CROSSFADE_DUR@).
crossfadeDurMs :: Int
crossfadeDurMs = 150

-- | The live overlay state. 'fxTime' is the clock's @now@ in ms (advanced by
-- the loop or the tests); 'fxStart' is the one-shot start time ('Nothing' at
-- rest); 'fxOpacity' is the cached eased opacity refreshed by 'fxTick'.
data Fx = Fx
  { fxTime :: !Int
  , fxStart :: !(Maybe Int)
  , fxOpacity :: !Float
  } deriving (Show, Eq)

-- | @Fx::new(Clock::fixed())@ — idle at time 0, opacity pinned at 1.0. The
-- production loop seeds the real wall time via 'fxTickAt' on the first frame.
fxNew :: Fx
fxNew = Fx{fxTime = 0, fxStart = Nothing, fxOpacity = 1.0}

-- | Advance the ramp to the clock's @now@ and cache the eased opacity. The
-- loop calls this once per frame (passing @clockNow@); tests use 'fxAdvance'
-- on a zero-start 'Fx'. Re-requests a frame while the fade is in flight (the
-- loop's idle pace already re-pumps, so this is a pure no-op marker here —
-- kept for API parity).
fxTickAt :: Int -> Fx -> Fx
fxTickAt now f =
  let (op, mStart) = opacityAt now f
  in f{fxTime = now, fxOpacity = op, fxStart = mStart}

-- | 'fxTickAt' at the current 'fxTime' (recompute opacity without moving
-- time). Mirrors @Fx::tick@ reading @clock.now()@.
fxTick :: Fx -> Fx
fxTick f = fxTickAt (fxTime f) f

-- | Advance the clock by @ms@ then 'fxTick'. The deterministic tests use this
-- with a zero-start 'Fx' (the real loop uses 'clockNow' + 'fxTickAt').
fxAdvance :: Int -> Fx -> Fx
fxAdvance ms f = fxTickAt (fxTime f + ms) f

-- | Re-trigger the crossfade (called on selection change). A no-op when
-- animations are disabled — BUT animations are an env-checked 'IO' concern,
-- so this is the UNGATED primitive; the on-event bridge checks
-- 'animationsEnabled' (IO) before calling it. Mirrors @retarget_crossfade@;
-- tests use 'retargetCrossfadeForce' directly to avoid the process-global
-- env race.
retargetCrossfade :: Fx -> Fx
retargetCrossfade = retargetCrossfadeForce

-- | Ungated crossfade re-trigger — set the one-shot start to the current
-- 'fxTime'. Mirrors @retarget_crossfade_force@.
retargetCrossfadeForce :: Fx -> Fx
retargetCrossfadeForce f = f{fxStart = Just (fxTime f)}

-- | Cached eased crossfade opacity (0.0..=1.0). Peek, no mutation.
crossfadeOpacity :: Fx -> Float
crossfadeOpacity = fxOpacity

-- | Eased opacity at @now@; clears the one-shot once 'crossfadeDurMs' elapses
-- (leaving opacity pinned at 1.0). Internal helper for 'fxTickAt'.
opacityAt :: Int -> Fx -> (Float, Maybe Int)
opacityAt now f = case fxStart f of
  Nothing -> (1.0, Nothing)
  Just start ->
    let elapsed = now - start
    in if elapsed >= crossfadeDurMs
         then (1.0, Nothing)
         else (clamp01 (tweenSample crossfadeTween elapsed), Just start)

-- | The 0→1 'EaseOut' 'Tween' over 'crossfadeDurMs'. Stateless, so sampled at
-- any @elapsed@ without per-instance state.
crossfadeTween :: Tween
crossfadeTween = tweenEasing EaseOut (tween 0.0 1.0 crossfadeDurMs)

-- | 'True' unless @DOTS_NO_ANIM@ is set. IO (reads the process env).
animationsEnabled :: IO Bool
animationsEnabled = do
  m <- lookupEnv "DOTS_NO_ANIM"
  pure (case m of Nothing -> True; Just _ -> False)

clamp01 :: Float -> Float
clamp01 x = max 0.0 (min 1.0 x)

-- * CrossfadeProbe

-- | Deterministic probe over the crossfade 'Tween', sampled at arbitrary
-- times (the 'Tween' is stateless, so call order does not matter). Used by
-- tests.
newtype CrossfadeProbe = CrossfadeProbe { probeNow :: Int -> Float }

-- | Eased 0→1 over @durMs@ with 'EaseOut'. The property tested: 0 at the
-- start, 1 at the duration, strictly between mid-fade. The 'Clock' argument
-- Rust takes is unused (@_clock@), so this port elides it.
crossfadeCurve :: Int -> CrossfadeProbe
crossfadeCurve durMs =
  CrossfadeProbe (clamp01 . tweenSample (tweenEasing EaseOut (tween 0.0 1.0 durMs)))

-- * blendBitmap

-- | Blend @bmp@ toward @bg@ by @opacity@ (0.0 → fully @bg@, 1.0 → fully
-- @bmp@), returning a fresh 'Bitmap'. Used by the view to render the
-- crossfade. NOTE: abstracttui's 'AbstractTUI.Gfx.Bitmap.blendBitmap' is
-- INVERTED (opacity 0→bmp, 1→bg), so this is a local definition with the
-- Rust orientation and half-away-from-zero channel rounding (matching Rust's
-- @f64::round@, NOT Haskell's 'round').
blendBitmap :: Bitmap -> Rgba -> Double -> Bitmap
blendBitmap bmp bg opacity =
  mapPixels (\p -> lerpAway (realToFrac opacity) bg p) bmp

-- | @lerp t a b = a + t*(b-a)@ with each channel rounded half away from zero
-- (Rust's @f64::round@) and clamped to 0..255. Alpha is lerped the same way
-- (the pane background is opaque, so it stays 255 in practice).
lerpAway :: Double -> Rgba -> Rgba -> Rgba
lerpAway t a b =
  Rgba
    { rgbaR = ch rgbaR
    , rgbaG = ch rgbaG
    , rgbaB = ch rgbaB
    , rgbaA = ch rgbaA
    }
  where
    t' = max 0.0 (min 1.0 t)
    ch f =
      let v = fromIntegral (f a) + t' * (fromIntegral (f b) - fromIntegral (f a)) :: Double
      in clampWord8 (roundHalfAway v)

-- | Clamp a rounded channel into the Word8 range.
clampWord8 :: Int -> Word8
clampWord8 n
  | n < 0 = 0
  | n > 255 = 255
  | otherwise = fromIntegral n